### PR TITLE
[Persistence][Utility] Proposal block metadata maintained in `utilityContext` rather than `postgresContext`

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

Maintain the proposed block in the `utilityContext` rather than the `postgresContext.`

This is a cleanup with some tech debt to help with the separation of concerns w.r.t contexts:
- Utility - has a semantic understanding of what is in the proposed block and whether it is valid
- Persistence - Responsible for computing & committing state integrity, but is agnostic to the actual state change of the transactions being proposed/applied.

This is a stepping stone to help with:
- Persistence rollback - changes to the SQL DB and Trees (i.e. KV stores) needs to be atomic and then either reverted or committed to disk
- Consensus state sync - when a node is catching up, it will also go through the `Propose` -> `Apply` -> `Commit` lifecycle, but it won't be dependant on the latency of consensus since it happened in the past so the `proposer` and `commitQC`  are available immediately

## Issue

Fixes part of #361:

Screenshot 2022-12-10 at 4 27 06 PM

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

## List of changes

**PersistenceModule Changes:**
- Remove `SetProposalBlock` and local vars to keep proposal state
- Add `proposerAddr` to the `Commit` function
- Move the `PostgresContext` struct to `context.db`

**Utility Module Changes:**
- Introduce `SetProposalBlock` and local vars to keep proposal state
- Maintaining proposal block state (proposer, hash, transactions) in local context

**Other Changes:**
- Partially start consolidating `appHash` and `stateHash` terminology
- Partially start consolidating all `hash` related vars to `string` instead of `[]byte`

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [x] I have updated the corresponding README(s); local and/or global
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [x] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: consensus/consensus_tests/utils_test.go
Comment (gokutheengineer) [inline consensus/consensus_tests/utils_test.go:377]: @Olshansk This causes tests to fail to compile, since struct `MockUtilityContext` doesn't implement function `SetProposalBlock`.

**Review by gokutheengineer — CHANGES_REQUESTED** [review]

File: consensus/consensus_tests/utils_test.go
Comment (Olshansk) [reply consensus/consensus_tests/utils_test.go:377]: Did this happen to you locally?

Note how the CI checks passed:

Screenshot 2022-12-13 at 4 44 43 PM

Here's also a screenshot of it passing on my local machine:

Screenshot 2022-12-13 at 4 45 11 PM

And in my `utility_module_mock.go`:

Screenshot 2022-12-13 at 4 46 40 PM


When we change the interfaces (and in turns the mocks on which the interfaces are built), we need to run `make mockgen`, so they are recompiled. I like to run `make develop_test` whenever I want to do a clean end-to-end.

Lmk if that makes sense

File: consensus/types/proto/block.proto
Comment (deblasis) [inline consensus/types/proto/block.proto:17]: Not actionable, only a "stupid question".

Just for my understanding: we are going to prefer strings to byte arrays for anything that's addresses and hashes, correct?

Is it for simplicity or there are other reasons?

File: persistence/test/block_test.go
Comment (deblasis) [inline persistence/test/block_test.go:13]: I am not sure if it's intended but as basically a newbie in this part of the code, I find it "odd" even if it makes sense to see

```go
stateHash, err := db.GetBlockHash()
```
shouldn't this be 

```go
blockHash, err := db.GetBlockHash()
```
?

File: persistence/genesis.go
Comment (deblasis) [inline persistence/genesis.go:148]: I would probably at least log the stateHash for debugging purposes.

Any thoughts against that?

**Review by deblasis — APPROVED** [review]
Solid work, just a couple of comments/questions

File: consensus/types/proto/block.proto
Comment (Olshansk) [reply consensus/types/proto/block.proto:17]: > Is it for simplicity or there are other reasons?

1. Simplicity - they're small, not big data blobs
2. Readability - Things like `address` or `stateHash` are "human readable" in the sense that we use & reference it (e.g. printing)
3. Performance (micro optimization) - lots of unnecessary encoding & decoding throughout the codebase

File: persistence/test/block_test.go
Comment (Olshansk) [reply persistence/test/block_test.go:13]: **tl;dr Yes**

Yea, the `appHash` and the `stateHash` are the same thing which was an inconsistency.

I still have to do a bit of research whether we are going to do anything different with `blockHash`. For example, the `quorumCert` and `proposer` are included in the block but not part of the state hash, but TBD.

Comment (Olshansk) [reply persistence/test/block_test.go]: Potentially addressed here: [`66ed7fc` (#391)](https://github.com/pokt-network/pocket/pull/391/commits/66ed7fc307771c9a57e1c31e21ee285984b471bf)

File: persistence/genesis.go
Comment (Olshansk) [reply persistence/genesis.go]: Potentially addressed here: [`39da0ed` (#391)](https://github.com/pokt-network/pocket/pull/391/commits/39da0ed1ab5e3dc8ac1101d2901fe8b752e5b7a2)

File: persistence/test/block_test.go
Comment (deblasis) [reply persistence/test/block_test.go]: Awesome, thank you! 🙏

**Comment by gokutheengineer** [comment]
As a general comment, and probably not in the scope of this issue, I think we need to divide UtilityContext interface to smaller interfaces, and only leave common functions that are to be used by both block proposers and block verifiers together. This will not change how utility operates, however it can maybe provide better code clarity. 

For example, `CreateAndApplyProposalBlock` is to be used by the block proposers, as mentioned in the comment:

``` go
	//Reaps the mempool for transactions to be proposed in a new block, and applies them to this
	//context; intended to be used by the block proposer.
	CreateAndApplyProposalBlock(proposer []byte, maxTransactionBytes int) (stateHash string, transactions [][]byte, err error)
```

Having a separate interface implemented by UtilityContext, such as for Proposer:

```go
type UtilityProposerContext interface {
       CreateAndApplyProposalBlock(proposer []byte, maxTransactionBytes int) (stateHash string, transactions [][]byte, err error)
}
```

might make the roles of Proposer and Verifier much more clear. In later stages we can consider, implementing only subset of  interfaces interfaces for different roles, and enforce strict control.

**Comment by Olshansk** [comment]
@gokutheengineer Completely agreed. Just a matter of prioritizing it. If you want to create a ticket or just add a TODO in the code, I'd 👍 it.

In #361, see the first action item:

Screenshot 2022-12-15 at 6 32 29 AM

I think this applies to the logic we have in utility, in persistence and in consensus too!

**Comment by Olshansk** [comment]
@gokutheengineer Also, since you requested change sin the past, this PR is blocked until you re-review / approve.

File: persistence/block.go
Comment (gokutheengineer) [inline persistence/block.go:38]: Similar to @deblasis's previous comment, out of curiosity, what is the main reason to move to string from bytes?

File: persistence/test/block_test.go
Comment (gokutheengineer) [reply persistence/test/block_test.go]: agreed, as my previous comment.

File: shared/modules/utility_module.go
Comment (gokutheengineer) [inline shared/modules/utility_module.go:29]: Shouldn't we add a comment stating this should be only used by the proposer? Or is there a reason for validators to run this?

**Comment by gokutheengineer** [comment]
Added few more comments on minor things. Please have a look, otherwise LGTM!

**Review by gokutheengineer — APPROVED** [review]

File: consensus/hotstuff_leader.go
Comment (Olshansk) [reply consensus/hotstuff_leader.go]: Adding this as a sub-bullet in https://github.com/pokt-network/pocket/issues/283#issue-1396979181 and updated the TODO in the code as well.

Comment (Olshansk) [reply consensus/hotstuff_leader.go]: I believe you're looking at an older sha of the PR, already done.

File: shared/modules/utility_module.go
Comment (Olshansk) [reply shared/modules/utility_module.go:29]: Adding the following comment:

```
	// This function is intended to be called by any type of node during state transitions. For example,
	// both block proposers and verifiers will use it to create a context (before finalizing it) during consensus,
	// and all verifiers will call it during state sync.
```

File: persistence/block.go
Comment (Olshansk) [reply persistence/block.go:38]: Same answer as here: https://github.com/pokt-network/pocket/pull/391#discussion_r1049075566

1. Simplicity - they're small, not big data blobs
2. Readability - Things like address or stateHash are "human readable" in the sense that we use & reference it (e.g. printing)
3. Performance (micro optimization) - lots of unnecessary encoding & decoding throughout the codebase

**Comment by Olshansk** [comment]
@gokutheengineer Even though I merged it in, please `Resolve` the comments or let me know if there's anything additional I can clarify.

**Pull request merged**
